### PR TITLE
33378 GraphQL: create an entry in 'customer_log' after generate customer token

### LINKED ISSUE #33378: Logging in or out as customer via graphql doesnt create a entry in 'customer_log'
### Preconditions (*)

1. Magento 2.4.2

### Steps to reproduce (*)

1. Log in as customer via graphql
2. Log out as customer via graphql

### Expected result (*)

1. An entry in the table 'customer_log' is created

### Actual result (*)

1.  No entry in table 'customer_log' is created

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by khrystynastolbova**
### Description (*)

Create an entry in 'customer_log' after generate customer token.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#33378

### Manual testing scenarios (*)

1. Check 'customer_log'  db table
2. Genarate customer token
```
mutation {
	generateCustomerToken(
        email: "roni_cost@example.com"
        password: "roni_cost3@example.com"
    ) {
        token
    }
}
```
3. Check 'customer_log' db table again. The field 'last_login_at' for this customer was updated.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by ihor-sviziev** [comment]
@magento run all tests

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ *  Customer log data logger test.
+ */
+class LoggerTest extends GraphQlAbstract
+{
+    /**
+     * Customer log data logger.
+     *
+     * @var Logger
+     */
+    protected $logger;
```
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php]:
```suggestion
    private $logger;
```

File: app/code/Magento/CustomerGraphQl/etc/graphql/events.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0"?>
+<!--
+/**
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+-->
+<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
```
Comment (eduard13) [inline app/code/Magento/CustomerGraphQl/etc/graphql/events.xml]:
```suggestion
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:noNamespaceSchemaLocation="urn:magento:framework:Event/etc/events.xsd">
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        $log = $this->logger->get(1);
+        $this->assertNotEmpty($log->getLastLoginAt());
+    }
+
+    /**
+     * @param string $email
+     * @param string $password
+     * @return string
+     */
+    private function getQuery(string $email = 'customer@example.com', string $password = 'password') : string
```
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php]:
```suggestion
    private function getQuery(string $email = 'customer@example.com', string $password = 'password'): string
```

**Review by eduard13 — CHANGES_REQUESTED** [review]
Thank you @khrystynastolbova for fixing this one. Could you please review the minor suggestions below?

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    {
+        $this->logger = Bootstrap::getObjectManager()->get(Logger::class);
+    }
+
+    /**
+     * Verify customer log after generate customer token
+     *
+     * @magentoApiDataFixture Magento/Customer/_files/customer.php
+     */
+    public function testGenerateCustomerValidToken()
```
Comment (ihor-sviziev) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/LoggerTest.php]:
```suggestion
    public function testGenerateCustomerValidToken(): void
```

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Here is one more improvement

**Comment by khrystynastolbova** [comment]
@ihor-sviziev and @eduard13, Thanks for your review and suggestions! I will make changes.
Maybe it would be better to move the logger test to GenerateCustomerTokenTest.php file and there add tearDown function to remove the customer log record? Because this test is failing
Screenshot 2021-07-05 at 12 16 34

**Comment by khrystynastolbova** [comment]
Please check my recent changes

**Comment by khrystynastolbova** [comment]
@magento run all tests

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/GenerateCustomerTokenTest.php [hunk 5]
```diff
@@ -152,25 +156,51 @@ public function testGenerateCustomerTokenWithEmptyEmail()
...
+     *
+     * @return void
+     */
+    protected function tearDown(): void
+    {
+        if ($this->logger->get(1)->getLastLoginAt()) {
+            /** @var AdapterInterface $connection */
+            /** @var ResourceConnection $resource */
+            $resource = Bootstrap::getObjectManager()->get(ResourceConnection::class);
+            $connection = $resource->getConnection(ResourceConnection::DEFAULT_CONNECTION);
```
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/GenerateCustomerTokenTest.php]:
```suggestion
            /** @var ResourceConnection $resource */
            $resource = Bootstrap::getObjectManager()->get(ResourceConnection::class);
            /** @var AdapterInterface $connection */
            $connection = $resource->getConnection(ResourceConnection::DEFAULT_CONNECTION);
```

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/GenerateCustomerTokenTest.php [hunk 6]
```diff
@@ -7,24 +7,41 @@
...
+use Magento\Framework\DB\Adapter\AdapterInterface;
+use Magento\TestFramework\Helper\Bootstrap;
 use Magento\TestFramework\TestCase\GraphQlAbstract;
 
 /**
  * API-functional tests cases for generateCustomerToken mutation
  */
 class GenerateCustomerTokenTest extends GraphQlAbstract
 {
+
```
Comment (eduard13) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Customer/GenerateCustomerTokenTest.php]:
```suggestion
```

**Review by eduard13 — CHANGES_REQUESTED** [review]
Good job. Could you apply these minor suggestions as well?

**Comment by eduard13** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by eduard13** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Review by eduard13 — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-9157](https://jira.corp.magento.com/browse/ENGCOM-9157) has been created to process this Pull Request

**Comment by eduard13** [comment]
@magento run Functional Tests B2B

**Comment by khrystynastolbova** [comment]
Hi @eduard13. Thanks for your review! I have applied your suggestions.

**Comment by eduard13** [comment]
@magento run all tests

**Review by eduard13 — CHANGES_REQUESTED** [review]
Hi @khrystynastolbova, could you please check the reason of failing Unit Tests builds?
Thank you.

**Comment by khrystynastolbova** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run all tests

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
@magento run Functional Tests CE

**Comment by engcom-Lima** [comment]
Hi @eduard13 the tests are now passed, moving this PR to Pending Review

**Review by eduard13 — APPROVED** [review]
Looks good for me 👍
Thanks for your input @khrystynastolbova

**Comment by magento-engcom-team** [comment]
Hi @eduard13, thank you for the review.
[ENGCOM-9157](https://jira.corp.magento.com/browse/ENGCOM-9157) has been created to process this Pull Request

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have a Magento lastest instance installed
2. Have one customer created for front end login as well


**Manual testing scenario:**

1. Make a note of customer last login time from _customer_log_ DB table (Customer id can be found in the _customer_entity_ table under _entity_id_ column).

2. Do get the customer token with GraphQL request
 
3. Re validate the customer last login is updated in the DB table.


**Before:** :heavy_multiplication_x: The last_login_at column data was not getting updated to the latest GraphQL login


**After:** :heavy_check_mark: The last_login_at column data was is getting updated to the latest GraphQL login


There is already a regression au-test has bee updated along with the PR changes, and hence no additional regression is required.

**Comment by ishakhsuvarov** [comment]
@magento run all tests

**Comment by vishaljpatil** [comment]
Hello @khrystynastolbova,

Can you please let us know about updating "last_logout_at" on customer revoke token using graphql (revokeCustomerToken).

On revoke token also customer log table "last_logout_at" column not updating.

**Pull request merged**